src/control/family.lean

```diff
@@ -0,0 +1,666 @@
+  (f' : α' ⟶ β') (g' : β' ⟶ γ') :
+  prod.map f f' ≫ prod.map g g' = prod.map (f ≫ g) (f' ≫ g') :=
+by apply prod.ext; simp
+
+@[simp]
+lemma prod.map_mk {α β α' β' : fam I}
+  (f :  α  ⟶ β) (g  : α'  ⟶ β') {i} (x : α i) (y : α' i) :
+  prod.map f g (x,y) = (f x,g y) :=
+rfl
+
+/-- diagonal arrow of the binary product in the category `fam I` -/
+def diag : Π {α : fam I}, α ⟶ α ⊗ α
+| α i x := (x,x)
+
+@[reassoc]
+lemma diag_map {α β : fam I} (f : α ⟶ β) : diag ≫ (f ⊗' f) = f ≫ diag :=
+by ext; refl
+
+@[reassoc]
+lemma diag_map_fst_snd {α β : fam I} : diag ≫ (prod.fst ⊗' prod.snd) = 𝟙 (α ⊗ β) :=
+by ext _ ⟨ ⟩; refl
+
+@[reassoc]
+lemma diag_map_comp {α β γ γ' : fam I} (f : α ⟶ β) (g : β ⟶ γ) (g' : β ⟶ γ') :
+  diag ≫ (f ≫ g ⊗' f ≫ g') = f ≫ diag ≫ (g ⊗' g') :=
+by ext; refl
+
+@[reassoc]
+lemma diag_map_fst_snd_comp {α β γ γ' : fam I} (g : α ⟶ γ) (g' : β ⟶ γ') :
+  diag ≫ (prod.fst ≫ g ⊗' prod.snd ≫ g') = (g ⊗' g') :=
+by ext _ ⟨ ⟩; refl
+
+/-- binary coproduct in the category `fam I` -/
+def sum (α β : fam I) : fam I
+| i := α i ⊕ β i
+
+localized "infix ` ⊕' `:35 := fam.sum" in fam
+
+/-- map function of the binary coproduct in the category `fam I` -/
+def sum.map {α β α' β' : fam I} : (α ⟶ β) → (α' ⟶ β') → (α ⊕' α' ⟶ β ⊕' β')
+| f g i (sum.inl x) := sum.inl $ f x
+| f g i (sum.inr x) := sum.inr $ g x
+
+localized "infix ` ⊕'' `:35 := fam.sum.map" in fam
+
+/-- left introduction arrow of the binary coproduct in the category `fam I` -/
+def sum.inl : Π {α β : fam I}, α ⟶ α ⊕' β
+| α β i x := _root_.sum.inl x
+
+/-- right introduction arrow of the binary coproduct in the category `fam I` -/
+def sum.inr : Π {α β : fam I}, β ⟶ α ⊕' β
+| α β i x := _root_.sum.inr x
+
+@[simp, reassoc]
+lemma sum.inl_map {α β α' β' : fam I} (f : α ⟶ β) (g : α' ⟶ β') :
+  sum.inl ≫ sum.map f g = f ≫ sum.inl :=
+by ext; refl
+
+@[simp, reassoc]
+lemma sum.inr_map {α β α' β' : fam I} (f : α ⟶ β) (g : α' ⟶ β') :
+  sum.inr ≫ sum.map f g = g ≫ sum.inr :=
+by ext; refl
+
+lemma sum.ext {α β α' : fam I} (f g : α ⊕' α' ⟶ β)
+  (h :  sum.inl ≫ f = sum.inl ≫ g)
+  (h' : sum.inr ≫ f = sum.inr ≫ g) :
+  f = g :=
+by { ext i (x|x),
+     apply congr_fun (congr_fun h  i) x,
+     apply congr_fun (congr_fun h' i) x }
+
+@[simp, reassoc]
+lemma sum.map_comp {α β γ α' β' γ' : fam I}
+  (f :  α  ⟶ β)  (g  : β  ⟶ γ)
+  (f' : α' ⟶ β') (g' : β' ⟶ γ') :
+  sum.map f f' ≫ sum.map g g' = sum.map (f ≫ g) (f' ≫ g') :=
+by apply sum.ext; simp
+
+/-- co-diagonal arrow of the binary coproduct in the category `fam I` -/
+def codiag : Π {α : fam I}, α ⊕' α ⟶ α
+| α i (_root_.sum.inl x) := x
+| α i (_root_.sum.inr x) := x
+
+end subtype
+
+open_locale fam
+
+@[simp]
+lemma comp_app {α β γ : fam I} (f : α ⟶ β) (g : β ⟶ γ) {i} (x : α i) : (f ≫ g) x = g (f x) := rfl
+
+/-- Propositional equality between values as a `Pred` -/
+protected def eq (α : fam I) : Pred (α ⊗ α) :=
+λ i x, x.1 = x.2
+
+/-- Application of predicate `p` to the target of arrow `f`. `f ⊨ p` is a proposition that
```

Comment:
I don't understand the comment or what the definition is doing. But maybe it is clear to others? It took me a while to figure out that this is equivalent to `∀ x j, p j (f x j)`, and then I realized that this is what `sat_intro` and `sat_elim` say. Why not use the simpler definition?

Reply:
It does seem that even mentioning `subtype` here is unnecessary.